### ISSUE TITLE
Desktop App

### ISSUE THREAD

**Issue by vHanda** [GitJournal/GitJournal#137]
It might be even easier to provide a desktop app of GitJournal over a web-app via [go-flutter](https://github.com/go-flutter-desktop/go-flutter).

**Comment by ataliba** [comment]
This will be amazing :-)

**Comment by vHanda** [comment]
<img width="528" alt="Screenshot 2020-05-31 at 09 00 09" src="https://user-images.githubusercontent.com/426467/83346495-2e09e180-a31d-11ea-870e-4ce637291ec6.png">

I got a basic version of the app running on osx. Lots of stuff to do in-order for this to work properly though -

- [ ] Replace libgit2 with dart_git
- [ ] Replace firebase analytics
- [ ] Remove Crashlytics
- [ ] Implement payment handling for Pro version
- [ ] Make the UI far more responsive. It's currently clearly a UI for a mobile.

**Comment by BWPanda** [comment]
So it sounds like this issue might deprecate https://github.com/GitJournal/GitJournal/issues/45 ...?

**Comment by vHanda** [comment]
> So it sounds like this issue might deprecate #45 ...?

Nah. I would like to have both Desktop and Web Version. The Web requires more work as we cannot use git directly or communicate via ssh. One would need to integrate with the GitHub and GitLab specific APIs.

**Comment by ericopnotes** [comment]
So I'll definitely need a web or desktop option, as it is really the only thing holding me back from using this as my "One true notebook". You have done some amazing work here @vHanda with GitJournal and if you could get even a beta Windows App out there I'd use it right away.

**Comment by vHanda** [comment]
> So I'll definitely need a web or desktop option, as it is really the only thing holding me back from using this as my "One true notebook". You have done some amazing work here @vHanda with GitJournal. Such a refined experience, like you actual are [dog-fooding](https://newrepublic.com/article/115349/dogfooding-tech-slang-working-out-glitches) this app yourself daily.

Thanks. I really appreciate your kind words. I've been a bit demotivated lately, so your comment really helps. I'll try to get the desktop version out as soon as I can.

**Comment by ataliba** [comment]
@vHanda your job is great. Make sure that even few users see it with a big difference in their daily lives.

**Comment by chris2fr** [comment]
Desktop App for which platform(s)? Since the connection to the GIT is necessary in realtime, maybe the web app is sufficient. Synchronization seems to be the issue for me.

**Comment by vHanda** [comment]
> Desktop App for which platform(s)? Since the connection to the GIT is necessary in realtime, maybe the web app is sufficient. Synchronization seems to be the issue for me.

All 3? Linux, Mac and Windows. I don't use Windows at all, so that will be a bit more challenging.

The issue with the web version is that it will only work with GitHub or Gitlab, and I will have to rewrite large parts to use their specific APIs, instead of normal SSH + git or even http + git.

Also, with the web-version, since GitHub and Gitlab don't have fine grained permissions, you will need to give GitJournal access to read+write to all your repos (public and private)

I'm the author of GitJournal, and I don't feel comfortable giving it access to all my repos. I wonder how others will feel.

**Comment by ericopnotes** [comment]
Yeah, I thought I noticed that the SSH key way had more targeted
permissions, not requiring unlimited API permission like the https style.
Is that true Vishesh?

**Comment by vHanda** [comment]
> Yeah, I thought I noticed that the SSH key way had more targeted permissions, not requiring unlimited API permission like the https style. Is that true Vishesh?

Classically Git can be accessed via 2 transport mechanisms -
* SSH - Fine grained permissions are possible via GitHub and GitLab
* HTTPS - Fine grained permissions also possible via GitHub and GitLab

However this https method cannot be easily used over the browser because of certain restrictions - You can read more about it over here - https://github.com/isomorphic-git/isomorphic-git#cors-support

Therefore in order to access Git over a browser, we need to use the custom APIs provided by Github and Gitlab. These custom APIs (as far as I can tell) do not allow fine grained permissions. It's an all or nothing  approach - [link](https://docs.github.com/en/developers/apps/scopes-for-oauth-apps).

GitHub (not GitLab) does have a newer concept called "GitHub Apps" which in theory can be granted lesser permissions than all or nothing - via the typical OAuth, but it's not clear to me if this would give per repository access. The [docs](https://docs.github.com/en/developers/apps/differences-between-github-apps-and-oauth-apps) aren't very clear.

Alternatively, there is always an option that users could create a new GitHub account and give that account write access to that one repo, and then give GitJournal full access to that alternative account. This sounds very cumbersome though.

**Comment by ericopnotes** [comment]
Yeah this is all stuff they ran into in trying to do it browser based on
https://github.com/taniarascia/takenote/issues/165 . Thanks for walking us
through it and providing links Vishesh 👍🕵️‍♀️👨‍🏫

**Comment by pfalcon** [comment]
> Desktop App

A humble opinion: for as long as GitJournal is a filesystem-based, Markdown-based note app with git integration, there's no need for "Desktop App", because for that, a user can use:

1. Any filesystem-based markdown note app, and there're many of them.
2. Any git client app, and there many of them (starting from the standard command-line git).

Please don't spread thin, please make a good mobile note-taking app, period. All other areas are overcrowded already.

**Comment by tionis** [comment]
@pfalcon Let me extend that:
Instead of developing a Desktop App, it would suffice (at least in my opinion) to write down a standard on how to handle notes in every detail. So other people could write for example a vscode plugin and call it a day. My reasoning for this is, that writing notes manually is possible but not desirable.

**Comment by vHanda** [comment]
Thanks for the feedback. I haven't worked much on the desktop app in a while. I won't be giving it much/any love for at least a few months. There are too many things todo on the mobile app.

@tionis : Could you please give me a rough example of what kind of standard you would like. it's not clear to me.

**Comment by tionis** [comment]
I just mean a markdown file defining the names used by files, the general file structure (including which files are picked up how), the yaml headers, what markdown parser is used and so forth

**Comment by BWPanda** [comment]
I've been waiting for a desktop app, and in the meantime have just been using GitHub's interface. However I have an issue where if I create a note on mobile, then edit it via GitHub, when I save it and access it on mobile again, the header is all mucked up. It seems that the way the metadata is stored isn't shown on mobile, but it is in GitHub, and then it's messed up on edits. Anyone know a way around this?

**Comment by vHanda** [comment]
@BWPanda : Could you please open a new issue with detailed instructions on how to reproduce it. (Maybe even screenshots) The more details the better. If I can reproduce it, I'll happily fix it.

**Comment by BWPanda** [comment]
@vHanda Hmmm, seems I can't reproduce this anymore... Maybe it was fixed since I last experienced it. In any case, if it happens again I'll open a new issue with STR.

**Comment by silvernode** [comment]
I made sure to search the issues before creating a new one and found exactly what I was looking for here. The main reason I would like a desktop app is that I would probably use GitJournal much more often with a physical keyboard. Sometimes I find myself wanting to write a journal entry but dread doing so on my phone with a touch screen keyboard. I went ahead and subscribed for $4.99 a month to help the project out but I think I am going to change it to $10.00 or more. I really really really! want a desktop app. It can be a flatpak on Linux or whatever is easiest to distribute with on each platform. 

Whatever I can do to make the desktop version a reality, please tell me and I will do my best to make it happen. Thank you all for this great project. I cannot express how much it has helped me. 

Cheers! Stay healthy and positive, and have a great day.

**Comment by vHanda** [comment]
Thanks @silvernode. I'm in the process of setting up some backend infra so I can process payments via Stripe, that'll mean not paying the 30% Google tax and it's one step closer to releasing the app on F-droid and creating the desktop app.

My plan is to have it by end Oct, but let's see.

Thank you for your awesome words. They are motivating.

**Comment by silvernode** [comment]
> 
> 
> Thanks @silvernode. I'm in the process of setting up some backend infra so I can process payments via Stripe, that'll mean not paying the 30% Google tax and it's one step closer to releasing the app on F-droid and creating the desktop app.
> 
> My plan is to have it by end Oct, but let's see.
> 
> Thank you for your awesome words. They are motivating.

Sounds great! You know, it might be fun for you to live stream a code sprint on the weekend sometime. I find it really helpful with motivation and other programmers tend to see things in the code that you might not notice. The part I really like is the getting everyone together to have a good time. Eventually you'll be able to take tips and stuff too. Live Streaming code sessions seems to be getting more and more popular as well. Keep up the good work.

**Comment by AdrienLeGuillou** [comment]
Following up on @pfalcon comment:

> @pfalcon Let me extend that:
> Instead of developing a Desktop App, it would suffice (at least in my opinion) to write down a standard on how to handle notes in every detail. So other people could write for example a vscode plugin and call it a day. My reasoning for this is, that writing notes manually is possible but not desirable.

I've been thinking about writing myself a neovim plugin for a while and having this kind of standard definition of what GitJournal needs/accepts would be amazing. I was also pondering whether writing a small python library for GitJournal could be usefull in this regard. I tend to switch editors quite often and such a library could simplify the process of building plugins for different editors. I will probably build this at some point and if you have thoughts on this I would enjoy everyones opinion.

**Comment by ericopnotes** [comment]
A vscode extension that could warn you if your syntax didn't matched the
GitJournal syntax would be appropriated. Also commands/actions like,
"GitJournal - start new checklist", that sort of thing, that puts down a
proper template to start that type of entry would be great.

To be honest I'm in vscode every day, searching projects and wrying
markdown, so even if the user experience isn't as slick as a dedicated
GitJournal desktop app, it would still be might handy.

**Comment by benji88i** [comment]
This is such a great app @vHanda .

I wanted to say how much I enjoy using this app because it's so simple and yet extremely powerful..

.I deleted my other note apps and got the pro version. 

I would love to see a Linux version. I use the desktop pc a lot more than my pixel 4. 

Currently running solid and fedora on my desktop and pop OS on my laptop's for work..

I am wondering if we are writing notes in "markdown mode", does that mean the file can be exported as a markdown file? 

Also what is metadata, and what is it used for in our notes? I write blogs actually sometimes too perhaps it helps...

**Comment by jrollins** [comment]
This may not be useful for what is being developed here, but [xotes](https://gitlab.com/jrollins/xotes) is a python library for interacting with gitjournal (that includes a full xapian index of all notes) which could easily be used as a backend for a desktop app (xotes comes with an ncurses interface).  If there's any interest in using it I'd be happy to help.

**Comment by pfalcon** [comment]
> but xotes

@jrollins, please consider a guthub mirror, the same way as you do with https://github.com/jrollins/xapers

**Comment by benji88i** [comment]
Hi sorry @jrollins ,

Was this directed to my comment about a desktop client? I am a bit lost about the software you mentioned if so

**Comment by jrollins** [comment]
@benji88i No, it was just a general comment to this issue the existence of a python library that supports gitjournal notes.  Xotes does work on linux though (which is my platform) so maybe you would find it useful.

**Comment by orelogo** [comment]
I recently started using GitJournal. Thank you @vHanda for the amazing product.!

I was looking for a desktop solution that didn't involve me having to manually run git commands. Then I realized that I can use GitHub itself as a web app for editing or creating new notes. This totally solves my use case. I think it would be worth advertising this on the product website.

**Comment by matthiasmiller** [comment]
Do you have any build instructions for Mac Desktop? I'm a developer who is very interested in this project and may be interested in contributing.

**Comment by jrollins** [comment]
> > [xotes](https://gitlab.com/jrollins/xotes) does handle auto-syncing when needed (e.g. before/after note editing).
> 
> I cannot get Xotes running. A terminal window opens and immediately disappears.

@thumos333 I'm sorry to hear that.  Maybe you can submit an issue report?  xotes does not try to open a terminal window, it's a command-line/ncurses app.  was there some kind of error message?

**Comment by thumos333** [comment]
> xotes does not try to open a terminal window, it's a command-line/ncurses app. was there some kind of error message?

@jrollins, thank you for replying. I tried to run it by making the .py file executable and double-clicking.

**Comment by jrollins** [comment]
> > xotes does not try to open a terminal window, it's a command-line/ncurses app. was there some kind of error message?
> 
> @jrollins, thank you for replying. I tried to run it by making the .py file executable and double-clicking.

ah yes I don't think that will work.  The application needs to be run in a terminal.  It currently doesn't have a gui, but it is fundamentally just a library, with a cli and curses interface, so it should be straightforward to make a gui similar to the curses interface.

**Comment by mikegr** [comment]
I tried all the recommended desktop apps, but I would say just out of convinience and familiarity with the mobile app, a desktop or web app would be really nice. It would much less mental work to use GitJournal on the desktop.

**Comment by vHanda** [comment]
I have good news. I've been using the desktop app for the past week. I can release a beta version soon. I've only used it on osx so far.

Also, it requires you to have "git" installed.

**Comment by ericopnotes** [comment]
Seems reasonable, and you can add basic instructions for installing git as
a pre-req, but I'm guessing your audience will 95%+ already have it
installed :).


This is exciting news! I'm definitely interested!

**Comment by glow12121** [comment]
> I tried all the recommended desktop apps, but I would say just out of convinience and familiarity with the mobile app, a desktop or web app would be really nice. It would much less mental work to use GitJournal on the desktop.

Good news.

**Comment by nth-chile** [comment]
I agree with @glow12121. For as many notetaking apps as there are, the only thing really missing is file syncing. I would prefer to choose my own desktop app (sublime text?, text edit?, github?) It shouldn't matter what the app is. The only thing I am missing is this: when I save, I want the note to sync across devices without overwriting. GitJournal is an attractive mobile app because it seems to understand that git or something similar is needed for proper sync to avoid overwriting. The missing piece, as @glow12121 is saying, is the part where, when I update my notes from my laptop (with any app), I want it to push the changes to git without running any commands or having to do any extra steps. CMD+S itself should be what triggers the git commands. I think what that means is, I want a file watcher.

There is a useful git script on the GJ docs [here](https://gitjournal.io/support/). But how do I get it to run on save? To me, THAT is the missing piece, not a desktop app. The reason that the GJ _mobile_ app is important is that consumers can't run scripts and file watchers from their phones. So, with a phone you need an app like GJ. With a desktop, you should be free to use any text editor, but also saving the file should sync

**Comment by vHanda** [comment]
Rough Update. Steps are missing for the desktop app -
- [ ] Creating ssh keys - My native dart implementation isn't reliable
- [ ] Cloning with those keys

(Both of these steps will involve calling a process on the desktop)

- [ ] Integrate the freshly implemented merge mechanism in Dart into GJ
- [ ] Figure out the packaging

@nth-chile : I agree with you. However, I think we should track this in another bug report. If it helps, I've already started working on an electron app which (for now) polls the folder. Could you please open another bug report? Once the basic scaffolding is in place, I can add watching.

PS: I know electron has its issues, but Flutter isn't mature enough for this right now.

If anyone wants to help, I would love a better icon. For me this is something independent of GitJournal, as it's just is just to sync to another git repo. The green folder is the temporary icon.

<img width="565" alt="Screenshot 2021-06-18 at 17 14 30" src="https://user-images.githubusercontent.com/426467/122582783-d5f16a80-d058-11eb-9411-7a927936e6a5.png">

**Comment by nth-chile** [comment]
> For me this is something independent of GitJournal, as it's just is just to sync to another git repo.

Awesome @vHanda, is it in a separate repo?

**Comment by vHanda** [comment]
> > For me this is something independent of GitJournal, as it's just is just to sync to another git repo.
> 
> Awesome @vHanda, is it in a separate repo?

I haven't pushed it so far, the code is todos and my experiments while I learn electron. Please just open the bug in this repo for now. It's just to create the Desktop App which polls / watches for changes.

**Comment by vHanda** [comment]
The final repo will be here - https://github.com/GitJournal/git-auto-sync . I've pushed some of the changes. It's just me experimenting right now. My local copy has a lot of uncommitted changes.

**Comment by nth-chile** [comment]
Created feature request https://github.com/GitJournal/GitJournal/issues/515

**Comment by vHanda** [comment]
Update: I never landed up figuring out the osx code signing, so we don't yet have a build you guys can use.

However, I did create an AppImage for Linux. It's built with each commit on the CI. Eg - https://github.com/GitJournal/GitJournal/runs/3191857379 (Look at the artifacts).

Please keep in mind that GitJournal needs a lot of UI work to make it more "responsive". Right now it clearly feels like a mobile app just running on the desktop.

**Comment by orbanbalage** [comment]
For macOS you will need an Apple Developer Account: https://learnappmaking.com/how-to-create-a-free-apple-developer-account/ (no need to pay for the AppStore stuff)

Then see here for example: https://wiki.lazarus.freepascal.org/Code_Signing_for_macOS#Using_codesign_to_sign_your_application

**Comment by vHanda** [comment]
Hi. I've done the basic codesigning for MacOS. However, I'm still figuring out the notarization, until then you'll have to explicitly allow GitJournal -

Eg - 

<img width="272" alt="Screenshot 2021-08-25 at 10 53 25" src="https://user-images.githubusercontent.com/426467/130761587-ec638b2f-6e1f-4db7-b67d-92397ccf781b.png">

You'll have to go to "Security and Privacy" and explicitly allow GitJournal.

<img width="694" alt="Screenshot 2021-08-25 at 11 02 30" src="https://user-images.githubusercontent.com/426467/130761697-e33476bb-c30d-4275-be76-1104044f2129.png">

You can get it from the CI. Eg - https://github.com/GitJournal/GitJournal/actions/runs/1165846131

**Comment by orbanbalage** [comment]
That's why I'm delaying the Big Sur update (viz. notarization)...

**Comment by ScriptAutomate** [comment]
Just want to say, it's really cool seeing the effort going into this project. Going with Flutter seems like an awesome choice, given their continued efforts to try and support Desktop, arm64, etc. as time goes on.

I picture a world where I can one day use GitJournal on a PinePhone :)
